packages/core/ui-settings/core-ui-settings-server-internal/src/clients/ui_settings_client.ts

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { UiSettingsClientCommon } from './ui_settings_client_common';
+import { BaseUiSettingsClient } from './base_ui_settings_client';
+import { UiSettingsServiceOptions } from '../types';
+
+interface UserProvidedValue<T = unknown> {
+  userValue?: T;
+  isOverridden?: boolean;
+}
+
+type UserProvided<T = unknown> = Record<string, UserProvidedValue<T>>;
+
+export class UiSettingsClient extends BaseUiSettingsClient {
+  private readonly uiSettingsClientCommon;
+
+  constructor(options: UiSettingsServiceOptions) {
+    const { log, defaults = {}, overrides = {} } = options;
+    super({ overrides, defaults, log });
+
+    this.uiSettingsClientCommon = new UiSettingsClientCommon(options);
```

Comment:
Can't we have `UiSettingsClientCommon extends BaseUiSettingsClient` and then `UiSettingsClient extends UiSettingsClientCommon`? I'm fine using a delegation pattern, but having a mix of delegation and inheritance feels wrong?

Reply:
I thought about this at length; my conclusion was that I intentionally want to avoid `UiSettingsClientCommon` to extend `BaseUiSettingsClient`; if it were to do so, it would mean that it can be instantiated in place of a `BaseUiSettingsClient`, and we don't want it to do so, as on its own it has no meaning.

Reply:
In the future we might want to add another instance of `BaseUiSettingsClient` that won't use any of the utility methods in `UiSettingsClientCommon`, and I think that's fine from OOP perspective? Happy to hear counter-arguments though.

Reply:
>  if it were to do so, it would mean that it can be instantiated in place of a BaseUiSettingsClient

Can't we just have `UiSettingsClientCommon` be abstract then, so that we're sure that it can't be instantiated directly (Or maybe we can't have an abstract class extends a non-abstract one? not sure with JS/TS)?

> and I think that's fine from OOP perspective?

Probably fine...ish 😄. As said, mixing inheritance / polymorphism and delegation feels strange. OTOH given the spaghetti code that this domain is, we're not making it worse anyway by doing that, so I don't have any major objection.

Reply:
This makes sense, I've made `UiSettingsClientCommon` to be abstract